src/perf/bin/drvmain.cpp

```diff
@@ -0,0 +1,604 @@
+    __analysis_assume(QueueConfig.EvtIoStop == 0);
+
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfIoQueueCreate failed");
+        goto Error;
+    }
+
+    QuicPerfCtlDevice = Device;
+    QuicPerfCtlExtension = DeviceContext;
+
+    WdfControlFinishInitializing(Device);
+
+    QuicTraceLogVerbose(
+        PerfControlInitialized,
+        "[perf] Control interface initialized");
+
+Error:
+
+    if (DeviceInit) {
+        WdfDeviceInitFree(DeviceInit);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+VOID
+QuicPerfCtlUninitialize(
+)
+{
+    QuicTraceLogVerbose(
+        PerfControlUninitializing,
+        "[perf] Control interface uninitializing");
+
+    if (QuicPerfCtlDevice != nullptr) {
+        NT_ASSERT(QuicPerfCtlExtension != nullptr);
+        QuicPerfCtlExtension = nullptr;
+
+        WdfObjectDelete(QuicPerfCtlDevice);
+        QuicPerfCtlDevice = nullptr;
+    }
+
+    QuicTraceLogVerbose(
+        PerfControlUninitialized,
+        "[perf] Control interface uninitialized");
+}
+
+PAGEDX
+_Use_decl_annotations_
+VOID
+QuicPerfCtlEvtFileCreate(
+    _In_ WDFDEVICE /* Device */,
+    _In_ WDFREQUEST Request,
+    _In_ WDFFILEOBJECT FileObject
+)
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    PAGED_CODE();
+
+    KeEnterGuardedRegion();
+    ExfAcquirePushLockExclusive(&QuicPerfCtlExtension->Lock);
+
+    do
+    {
+        if (QuicPerfCtlExtension->ClientListSize >= 1) {
+            QuicTraceEvent(
+                LibraryError,
+                "[ lib] ERROR, %s.",
+                "Already have max clients");
+            Status = STATUS_TOO_MANY_SESSIONS;
+            break;
+        }
+
+        QUIC_TEST_CLIENT* Client = QuicPerfCtlGetFileContext(FileObject);
+        if (Client == nullptr) {
+            QuicTraceEvent(
+                LibraryError,
+                "[ lib] ERROR, %s.",
+                "nullptr File context in FileCreate");
+            Status = STATUS_INVALID_PARAMETER;
+            break;
+        }
+
+        RtlZeroMemory(Client, sizeof(QUIC_TEST_CLIENT));
+
+        //
+        // Insert into the client list
+        //
+        InsertTailList(&QuicPerfCtlExtension->ClientList, &Client->Link);
+        QuicPerfCtlExtension->ClientListSize++;
+
+        QuicTraceLogInfo(
+            TestControlClientCreated,
+            "[test] Client %p created",
+            Client);
+
+        //
+        // Update globals. (TODO: Add multiple device client support)
+        //
+        QuicPerfClient = Client;
+    } while (false);
+
+    ExfReleasePushLockExclusive(&QuicPerfCtlExtension->Lock);
+    KeLeaveGuardedRegion();
+
+    WdfRequestComplete(Request, Status);
+}
+
+PAGEDX
+_Use_decl_annotations_
+VOID
+QuicPerfCtlEvtFileClose(
+    _In_ WDFFILEOBJECT /* FileObject */
+)
+{
+    PAGED_CODE();
+}
+
+PAGEDX
+_Use_decl_annotations_
+VOID
+QuicPerfCtlEvtFileCleanup(
```

Comment:
Once you put the stop event in the client struct, you should set the event in here and the IoCanceled function below.